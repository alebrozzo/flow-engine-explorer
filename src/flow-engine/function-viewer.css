.flow_function__container {
    display: flex;
    flex-direction: column;
    border-radius: 0.25rem;
    position: relative;
    box-shadow: 0 0 1.5rem 0 rgba(72, 94, 116, 0.7);
    background-color: white;
    margin: 2rem;
    width: 40rem;
}

.flow_function__header {
    width: 100%;
    cursor: pointer;
    background: -ms-linear-gradient(125deg, rgb(192, 207, 228) 0%, rgb(9, 106, 151) 100%); /* ie10+ */
    background: linear-gradient(325deg, rgb(192, 207, 228) 0%, rgb(9, 106, 151) 100%);
}

.flow_function__header__id {
    display: inline-block;
    text-align: center;
    width: 3rem;
}

.flow_function__body {
    padding: 1rem;
}

.flow_function__body__code {
    font-family: "Liberation Mono", Menlo, Courier, monospace;
    white-space: pre-wrap;
    margin: 1rem 0;
    padding: 0.5rem;
    background-color: #fcfcfc;
}

.flow_function__body__next_group {
    display: flex;
    justify-content: space-between;
}

.flow_function__body__next {
    text-align: center;
    margin-top: 0.25rem;
}

.flow_function__body__next_pass {
    padding: 1rem;
    background-color: rgba(0, 255, 0, 0.2);
}

.flow_function__body__next_fail {
    padding: 1rem;
    background-color: rgba(255, 0, 0, 0.2);
}

@media only screen and (max-width: 700px) {
    .flow_function__container {
        width: 400px;
    }
}
