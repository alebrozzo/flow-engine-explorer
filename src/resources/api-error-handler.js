// @flow
export function normalizeApiErrors(err: any, logToConsole?: boolean): Array<string> {
    if (logToConsole) {
        console.error("API error:", err);
    }
    let errors: Array<string> = [];
    const statusCode = (err.response && err.response.status) || err.status || -1;
    if (statusCode === 500) {
        errors.push("There was an error accessing the server.");
    } else if (statusCode === 401 || statusCode === 403) {
        errors.push(
            "You do not have permissions to access this feature.\nPlease contact an Administrator for assistance."
        );
    } else if (statusCode === 404) {
        errors.push("The resource was not found.");
    } else if (statusCode === 400 && err.response && err.response.data) {
        Object.keys(err.response.data).forEach(key => {
            errors = errors.concat(err.response.data[key]);
        });
    } else {
        errors.push(err.message || "Unknown error.");
    }
    if (logToConsole) {
        console.error("API error list:", errors);
    }
    return errors;
}
