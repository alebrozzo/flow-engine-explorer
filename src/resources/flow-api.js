export const getFlowFunctions = () =>
    // simulate an API hit delay of a random time between 0.2 and 1 seconds
    new Promise((resolve, reject) => {
        setTimeout(() => resolve(getData()), Math.floor(Math.random() * 800) + 200);
    });

export const getData = () => [
    {
        id: 1,
        title: "Verify if weekend",
        body: `function isWeekend(date) {
  const dayOfWeek = date.getDay();
  return (dayOfWeek === 6) || (dayOfWeek === 0);
}`,
        true_id: 2,
        false_id: 3,
    },
    {
        id: 2,
        title: "Celebrate",
        body: `function celebrate() {
  console.log('Party on!!!');
  return true;
}`,
        true_id: null,
        false_id: null,
    },
    {
        id: 3,
        title: "Verify if Friday",
        body: `function isItFriday(date) {
  const dayOfWeek = date.getDay();
  return dayOfWeek === 5;
}`,
        true_id: 4,
        false_id: 5,
    },
    {
        id: 4,
        title: "TGIF",
        body: `function tgif() {
  console.log('TGIF!');
  return true;
}`,
        true_id: null,
        false_id: null,
    },
    {
        id: 5,
        title: "Groan",
        body: `function groan(date) {
  const daysUntilWeekend = 6 - date.getDay();
  console.log('Hang in there... ' + daysUntilWeekend + ' days until the weekend!');
  return true;
}`,
        true_id: null,
        false_id: null,
    },
];
