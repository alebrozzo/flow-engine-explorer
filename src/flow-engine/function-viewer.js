// @flow
import * as React from "react";

import { type FlowFunction } from "../types/flow-function";
import "./function-viewer.css";

type Props = {
    flowFunction: FlowFunction,
    passed: boolean,
};

type State = {
    isOpen: boolean,
};

export class FunctionViewer extends React.PureComponent<Props, State> {
    state: State = {
        isOpen: false,
    };

    toggleView = () => {
        this.setState({ isOpen: !this.state.isOpen });
    };

    render() {
        const { isOpen } = this.state;
        const { flowFunction, passed } = this.props;

        return (
            <div className="flow_function__container">
                <div
                    className={"flow_function__header flow_function__header-" + (passed ? "pass" : "fail")}
                    onClick={this.toggleView}>
                    <h2>
                        <span className="flow_function__header__id">{flowFunction.id}</span>
                        {flowFunction.title}
                    </h2>
                </div>
                <div className="flow_function__body">
                    {isOpen && (
                        <React.Fragment>
                            <label htmlFor={`flowBody_${flowFunction.id}`}>Rule body</label>
                            <div className="flow_function__body__code" id={`flowBody_${flowFunction.id}`}>
                                {flowFunction.body}
                            </div>
                        </React.Fragment>
                    )}
                    <div className="flow_function__body__next_group">
                        <div className="flow_function__body__next_pass">
                            <label htmlFor={`flowPass_${flowFunction.id}`}>Next if passed</label>
                            <div className="flow_function__body__next" id={`flowPass_${flowFunction.id}`}>
                                {flowFunction.true_id || "End"}
                            </div>
                        </div>
                        <div className="flow_function__body__next_fail">
                            <label htmlFor={`flowPass_${flowFunction.id}`}>Next if failed</label>
                            <div className="flow_function__body__next" id={`flowPass_${flowFunction.id}`}>
                                {flowFunction.false_id || "End"}
                            </div>
                        </div>
                    </div>
                </div>
            </div>
        );
    }
}
