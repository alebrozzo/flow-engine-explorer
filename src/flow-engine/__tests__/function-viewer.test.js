import React from "react";
import { shallow } from "enzyme";
import toJson from "enzyme-to-json";

import { FunctionViewer } from "../function-viewer";
import { getData } from "../../resources/flow-api";

describe("FunctionViewer", () => {
    let defaults;
    beforeAll(() => {
        defaults = {
            functions: getData(),
        };
    });

    it("1 - renders a function's data", () => {
        const component = shallow(<FunctionViewer flowFunction={defaults.functions[0]} passed={true} />);
        expect(toJson(component)).toMatchSnapshot();
    });

    it("2 - renders a function's data when no more functions in chain", () => {
        const fn = getData()[1];
        const component = shallow(<FunctionViewer flowFunction={defaults.functions[1]} passed={true} />);
        expect(toJson(component)).toMatchSnapshot();
    });

    it("3 - renders pass header properly", () => {
        const fn = getData()[1];
        const component = shallow(<FunctionViewer flowFunction={defaults.functions[1]} passed={true} />);
        expect(toJson(component)).toMatchSnapshot();
    });

    it("4 - renders fail header properly", () => {
        const fn = getData()[1];
        const component = shallow(<FunctionViewer flowFunction={defaults.functions[1]} passed={false} />);
        expect(toJson(component)).toMatchSnapshot();
    });
});
