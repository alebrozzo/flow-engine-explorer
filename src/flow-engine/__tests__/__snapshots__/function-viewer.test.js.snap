// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FunctionViewer 1 - renders a function's data 1`] = `
<div
  className="flow_function__container"
>
  <div
    className="flow_function__header flow_function__header-pass"
    onClick={[Function]}
  >
    <h2>
      <span
        className="flow_function__header__id"
      >
        1
      </span>
      Verify if weekend
    </h2>
  </div>
  <div
    className="flow_function__body"
  >
    <div
      className="flow_function__body__next_group"
    >
      <div
        className="flow_function__body__next_pass"
      >
        <label
          htmlFor="flowPass_1"
        >
          Next if passed
        </label>
        <div
          className="flow_function__body__next"
          id="flowPass_1"
        >
          2
        </div>
      </div>
      <div
        className="flow_function__body__next_fail"
      >
        <label
          htmlFor="flowPass_1"
        >
          Next if failed
        </label>
        <div
          className="flow_function__body__next"
          id="flowPass_1"
        >
          5
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`FunctionViewer 2 - renders a function's data when no more functions in chain 1`] = `
<div
  className="flow_function__container"
>
  <div
    className="flow_function__header flow_function__header-pass"
    onClick={[Function]}
  >
    <h2>
      <span
        className="flow_function__header__id"
      >
        2
      </span>
      Celebrate
    </h2>
  </div>
  <div
    className="flow_function__body"
  >
    <div
      className="flow_function__body__next_group"
    >
      <div
        className="flow_function__body__next_pass"
      >
        <label
          htmlFor="flowPass_2"
        >
          Next if passed
        </label>
        <div
          className="flow_function__body__next"
          id="flowPass_2"
        >
          End
        </div>
      </div>
      <div
        className="flow_function__body__next_fail"
      >
        <label
          htmlFor="flowPass_2"
        >
          Next if failed
        </label>
        <div
          className="flow_function__body__next"
          id="flowPass_2"
        >
          End
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`FunctionViewer 3 - renders pass header properly 1`] = `
<div
  className="flow_function__container"
>
  <div
    className="flow_function__header flow_function__header-pass"
    onClick={[Function]}
  >
    <h2>
      <span
        className="flow_function__header__id"
      >
        2
      </span>
      Celebrate
    </h2>
  </div>
  <div
    className="flow_function__body"
  >
    <div
      className="flow_function__body__next_group"
    >
      <div
        className="flow_function__body__next_pass"
      >
        <label
          htmlFor="flowPass_2"
        >
          Next if passed
        </label>
        <div
          className="flow_function__body__next"
          id="flowPass_2"
        >
          End
        </div>
      </div>
      <div
        className="flow_function__body__next_fail"
      >
        <label
          htmlFor="flowPass_2"
        >
          Next if failed
        </label>
        <div
          className="flow_function__body__next"
          id="flowPass_2"
        >
          End
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`FunctionViewer 4 - renders fail header properly 1`] = `
<div
  className="flow_function__container"
>
  <div
    className="flow_function__header flow_function__header-fail"
    onClick={[Function]}
  >
    <h2>
      <span
        className="flow_function__header__id"
      >
        2
      </span>
      Celebrate
    </h2>
  </div>
  <div
    className="flow_function__body"
  >
    <div
      className="flow_function__body__next_group"
    >
      <div
        className="flow_function__body__next_pass"
      >
        <label
          htmlFor="flowPass_2"
        >
          Next if passed
        </label>
        <div
          className="flow_function__body__next"
          id="flowPass_2"
        >
          End
        </div>
      </div>
      <div
        className="flow_function__body__next_fail"
      >
        <label
          htmlFor="flowPass_2"
        >
          Next if failed
        </label>
        <div
          className="flow_function__body__next"
          id="flowPass_2"
        >
          End
        </div>
      </div>
    </div>
  </div>
</div>
`;
